src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -427,18 +427,61 @@ void BindRangeType(NpgsqlTypeHandler elementHandler, PostgresRangeType pgRangeTy
 
         #endregion Binding
 
-        internal NpgsqlDbType GetNpgsqlTypeByOid(uint oid)
+        internal (NpgsqlDbType? npgsqlDbType, PostgresType postgresType, string dataTypeName) GetTypeInfoByOid(uint oid)
         {
-            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var pgType))
+            if (!DatabaseInfo.ByOID.TryGetValue(oid, out var postgresType))
                 throw new InvalidOperationException($"Couldn't find PostgreSQL type with OID {oid}");
 
-            if (!Mappings.TryGetValue(pgType.Name, out var mapping) && !Mappings.TryGetValue(pgType.FullName, out mapping))
-                throw new InvalidOperationException($"No mapping found for PostgreSQL type {pgType.DisplayName}");
+            // Try to find the postgresType in the mappings 
+            if (TryGetTypeInfoFromMapping(postgresType, out NpgsqlDbType? npgsqlDbType))
+                return (npgsqlDbType, postgresType, postgresType.DisplayName);
 
-            if (!mapping.NpgsqlDbType.HasValue)
-                throw new InvalidOperationException($"Invalid parameter type: {oid}");
+            if (postgresType is PostgresArrayType arrayType)
+            {
+                // Try to find the Elements' postgresType in the mappings 
+                if (TryGetTypeInfoFromMapping(arrayType.Element, out NpgsqlDbType? elementType))
+                    return (elementType | NpgsqlDbType.Array, postgresType, postgresType.DisplayName);
+
+                // It might be an array of an unmapped enum or composite type
+                if (arrayType.Element is PostgresEnumType || arrayType.Element is PostgresCompositeType)
+                    return (null, postgresType, postgresType.DisplayName);
+            }
+
+            // It might be an unmapped enum type
+            if (postgresType is PostgresEnumType || postgresType is PostgresCompositeType)
+                return (null, postgresType, postgresType.DisplayName);
+
+            throw new InvalidOperationException($"No mapping found for PostgreSQL type {postgresType.DisplayName}");
+        }
+
+        bool TryGetTypeInfoFromMapping(PostgresType pgType, out NpgsqlDbType? type)
```

Comment:
I'd recommend simplifying this and maybe changing the relationship between the two methods (`TryGetTypeInfoFromMapping` and `GetTypeInfoByOid)`:
* Enum/Composite is spread across the two: `TryGetTypeInfoFromMapping`checks for the factory type and returns true+null, which then makes `GetTypeInfoByOid` go ahead and check for `Postgres{Enum,Composite}Type`.
* NpgsqlDbType is determined by `TryGetTypeInfoFromMapping`, except for the array case where it's constructed by `GetTypeInfoByOid`.
* See comment below - the assumption seems to be that we can handle types with NpgsqlDbType and enums/composites, but other types without NpgsqlDbType aren't handled (exception thrown). I don't think this is correct - it's perfectly fine for mappings to not have an NpgsqlDbType (external plugins). Once NpgsqlDbType becomes optional, I think these two methods may change enough to warrant refactoring them into one method - but not sure.
* `type` here is an ambiguous name, should be `npgsqlDbType`

Do you want to take another look at this? I can also try to come up with a proposal if you want.

Reply:
Agreed and fixed.
`TryGetTypeInfoFromMapping` is now a tiny local helper function `TryGetMapping`